### PR TITLE
LP / QP solvers with vpLinProg and vpQuadProg

### PR DESCRIPTION
**Pull request by oKermorgant**
Small examples are available in the documentation.
More complete examples of QP have been added to show the performances.

### CODE REVIEW

File: example/math/qp_plot.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+vpColVector randV(int n)
+{
+  vpColVector M(n);
+  for(int i = 0; i < n; ++i)
+    M[i] = (2.*rand())/RAND_MAX -1;
+  return M;
+}
+
+#ifdef VISP_HAVE_DISPLAY
+class QPPlot
```
Comment (fspindle) [inline example/math/qp_plot.h]: `class VISP_EXPORT QPlot`

File: example/math/qp_plot.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+  void wait()
+  {
+    P->I.display->getClick();
+  }
+  vpPlot* P;
+};
+#else
+class QPPlot
+{
```
Comment (fspindle) [inline example/math/qp_plot.h]: `class VISP_EXPORT QPlot`

File: modules/core/include/visp3/core/vpLinProg.h [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        return false;
+    }
+    return true;
+  }
+  //@}
+};
+
+
+
+#endif // vpLinProgh
```
Comment (fspindle) [inline modules/core/include/visp3/core/vpLinProg.h]: Add an empty line at the end of each file

File: modules/core/src/tools/optimization/vpLinProg.cpp [hunk 4]
```diff
@@ -0,0 +1,770 @@
...
+ *
+ * Authors:
+ * Olivier Kermorgant
+ *
+ *****************************************************************************/
+
+#include <visp3/core/vpLinProg.h>
+
+using namespace std;
+
```
Comment (fspindle) [inline modules/core/src/tools/optimization/vpLinProg.cpp]: Please remove `using namespace std;`

File: modules/core/src/tools/optimization/vpLinProg.cpp [hunk 5]
```diff
@@ -0,0 +1,770 @@
...
+  }
+
+  std::vector<std::pair<double, unsigned int>> a;
+  a.reserve(n);
+
+  vpColVector R(n-m), db(m), dn(n-m);
+
+  while(true)
+  {
+
```
Comment (fspindle) [inline modules/core/src/tools/optimization/vpLinProg.cpp]: Remove empty line

File: modules/core/src/tools/optimization/vpLinProg.cpp [hunk 6]
```diff
@@ -0,0 +1,770 @@
...
+    for(unsigned int k = 0; k < m; ++k)
+    {
+      if(db[k] < -tol)
+        a.push_back({-x[B[k]]/db[k], k});
+    }
+    // get smallest index for smallest alpha
+    const auto amin = std::min_element(a.begin(), a.end());
+    const double alpha = amin->first;
+    const unsigned int k = amin->second;
+
```
Comment (fspindle) [inline modules/core/src/tools/optimization/vpLinProg.cpp]: Remove 1 empty line

File: modules/core/src/tools/optimization/vpLinProg.cpp [hunk 7]
```diff
@@ -0,0 +1,770 @@
...
+    // update A and c
+    std::swap(cb[k], cn[j]);
+    for(unsigned int i = 0; i < m; ++i)
+      std::swap(Ab[i][k], An[i][j]);
+
+    // update B and N
+    std::swap(B[k],N[j]);
+  }
+
+}
```
Comment (fspindle) [inline modules/core/src/tools/optimization/vpLinProg.cpp]: Add empty line at the end of the file

File: modules/core/src/tools/optimization/vpQuadProg.cpp [hunk 8]
```diff
@@ -0,0 +1,718 @@
...
+
+  Q = D*P*V.transpose();
+  r = -Q.t().pseudoInverse()*c;
+#else
+  throw(vpException(vpException::functionNotImplementedError, "Symmetric matrix decomposition is not implemented. You "
+                                                              "should install GSL 3rd party"));
+#endif
+}
+
+
```
Comment (fspindle) [inline modules/core/src/tools/optimization/vpQuadProg.cpp]: Remove 2 empty lines

File: modules/core/src/tools/optimization/vpQuadProg.cpp [hunk 9]
```diff
@@ -0,0 +1,718 @@
...
+  x1 = b;
+  Z = A;
+  if(A.getRows() == b.getRows() && vpLinProg::colReduction(Z, x1, false, tol))
+    return true;
+
+  std::cout << "vpQuadProg::setEqualityConstraint: equality constraint infeasible" << endl;
+  return false;
+}
+
+
```
Comment (fspindle) [inline modules/core/src/tools/optimization/vpQuadProg.cpp]: Remove 1 empty line

File: modules/core/src/tools/optimization/vpQuadProg.cpp [hunk 10]
```diff
@@ -0,0 +1,718 @@
...
+/*!
+  Solves a Quadratic Program under equality constraints.
+
```
Comment (fspindle) [inline modules/core/src/tools/optimization/vpQuadProg.cpp]: remove 1 empty line

File: modules/core/src/tools/optimization/vpQuadProg.cpp [hunk 11]
```diff
@@ -0,0 +1,718 @@
...
+  else if(vpLinProg::allLesser(C, b, d, tol))  // Ax = b has only 1 solution
+  {
+    x = b;
+    return true;
+  }
+  std::cout << "vpQuadProg::solveQP: inequality constraint infeasible" << endl;
+  return false;
+}
+
+
```
Comment (fspindle) [inline modules/core/src/tools/optimization/vpQuadProg.cpp]: Remove 1 empty line

File: modules/core/src/tools/optimization/vpQuadProg.cpp [hunk 12]
```diff
@@ -0,0 +1,718 @@
...
+      }
+      else
+        update_Ap = false;
+      // update x for next iteration
+      x += alpha * u;
+      g -= alpha*Q*u;
+    }
+  }
+  return true;
+}
```
Comment (fspindle) [inline modules/core/src/tools/optimization/vpQuadProg.cpp]: Add empty line at the end of the file

**Comment by fspindle** [comment]
@oKermorgant Can I merge, or do you want to introduce other changes ?

**Comment by oKermorgant** [comment]
That is fine for me!

**Pull request merged**